### train ANN Model



In [1]:
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.models import Model

In [2]:
from sklearn.model_selection import train_test_split

In [3]:
data = pd.read_csv("data_dropped.csv")

In [4]:
data.columns

Index(['Category', 'url_len', '@', '?', '-', '=', '.', '#', '%', '+', '$', '!',
       '*', ',', '//', 'abnormal_url', 'https', 'digits', 'letters',
       'Shortining_Service', 'having_ip_address'],
      dtype='object')

## data

In [5]:
x_train, x_test, y_train, y_test = train_test_split(data.iloc[:,1:], data.iloc[:,0])

In [6]:
inputs = Input(shape=(20,))

x = Dense(128, activation='sigmoid')(inputs)
x = Dense(64, activation='sigmoid')(x)

predictions = Dense(2, activation='softmax')(x)

mlp = Model(inputs=inputs, outputs=predictions)

In [7]:
mlp.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])

In [8]:
mlp.fit(x_train,y_train,epochs=10)

Epoch 1/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 10s 638us/step - accuracy: 0.8525 - loss: 0.3417
Epoch 2/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 689us/step - accuracy: 0.8927 - loss: 0.2802
Epoch 3/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 699us/step - accuracy: 0.9125 - loss: 0.2295
Epoch 4/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 699us/step - accuracy: 0.9172 - loss: 0.2188
Epoch 5/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 700us/step - accuracy: 0.9189 - loss: 0.2145
Epoch 6/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 707us/step - accuracy: 0.9208 - loss: 0.2095
Epoch 7/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 710us/step - accuracy: 0.9214 - loss: 0.2059
Epoch 8/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 712us/step - accuracy: 0.9215 - loss: 0.2043
Epoch 9/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 721us/step - accuracy: 0.9226 - loss: 0.1998
Epoch 10/10
15263/15263 ━━━━━━━━━━━━━━━━━━━━ 11s 718us/step - accuracy: 0.9231 - loss: 0.1988


In [10]:
import numpy as np
y_pred = np.argmax(mlp.predict(x_test), axis=-1)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)

5088/5088 ━━━━━━━━━━━━━━━━━━━━ 3s 631us/step
Accuracy: 0.9212828167422205
Precision: 0.9649705014749262
Recall: 0.7988830828350303
F1 Score: 0.8741072568840687


### Test Result
With 2 layers of perceptron (128, 64), 10 epochs, both with Sigmoid function, we got : 


Accuracy: 0.9212828167422205

Precision: 0.9649705014749262

Recall: 0.7988830828350303

F1 Score: 0.8741072568840687